deepspeed/runtime/config.py

```diff
@@ -211,8 +211,21 @@ def get_zero_reduce_scatter(param_dict):
                             ZERO_OPTIMIZATION_REDUCE_SCATTER_DEFAULT)
 
 
-def get_allreduce_always_fp32(param_dict):
-    return get_scalar_param(param_dict, FP32_ALLREDUCE, FP32_ALLREDUCE_DEFAULT)
+def get_communication_data_type(param_dict):
+    val = get_scalar_param(param_dict,
+                           COMMUNICATION_DATA_TYPE,
+                           COMMUNICATION_DATA_TYPE_DEFAULT)
+    val = val.lower()
+    if val == "none":

```

Comment:
I can't think of a situation where a none value for communication_data_type is useful. I think we should have fp16 as the default value. Can you share your thoughts here?

Reply:
In this case, `none` means do not change the model's data type. Perhaps we could rename `none` to `default` for more readability. I think it is a bad idea to forcefully set fp16 data type for communication because it is can reduce the quality of the model if we train the model in fp32 mode. I think it's better for the user to set these values himself.

Reply:
Yes, you are absolutely correct; I was wrong. Using the model data type as the default communication data type is the right thing to do. However, I am unsure that `default` actually improves readability since this option is meant to help users override the default communication data type. So I am wondering if a better approach is for users to only use this option when they want to override since deepspeed can internally figure out the model data type in the default case. In other words the only valid values for this configuration option will be override data types, e.g., fp16, fp32, etc. What do you think?

Reply:
So your suggestion is to communicate between GPUs with the `float16` as the default data type?
I think this is based on how the project owners see the main purpose of this project. If the main purpose is the training speed that's ok. If the main purpose is speed without reducing the model quality, I think this is not the best option.
In any case, if we decide to use fp16 as the default communication data type, we need to properly describe this in the documentation. This is where I need your help to point me to where I need to add information about these changes to the documentation.
Another problem with these changes is that when users upgrade the package to the version with these changes, their results may differ from their old results.
So if your think using fp16 as a default communication data type is a good idea - I will remove `default` and will only use torch data types as valid arguments.

Reply:
No, I don't think fp16 should be default communication data type. I want the model training data type (e.g., fp32, fp16, bfloat16) to be the default communication data type, and the configuration option `communication_data_type` should be a way for users to override the communication data type. 

I am sorry that I am not explaining clearly. Please don't make any more changes yet, as you have already done a lot of work. I will send a better explanation when I have a bit more time. Sorry about this.

Reply:
ok, no problem. I think I understand what you mean. Just remove `default` as a valid argument and set `COMMUNICATION_DATA_TYPE_DEFAULT=None`. And when`communication_data_type=None` set it is to a valid torch data type in `engine.py` when model and all arguments will be set.

Reply:
Yes, we are aligned. Thanks so much for your help.